src/dmd/dtemplate.d

```diff
@@ -5808,6 +5803,41 @@ extern (C++) class TemplateInstance : ScopeDsymbol
     TemplateInstance tnext;     // non-first instantiated instances
     Module minst;               // the top module that instantiated this instance
 
+    ushort nest;                // for recursive pretty printing detection, 3 MSBs reserved for flags (below)
+    ubyte inuse;                // for recursive expansion detection
+
+    private enum Flag : uint
+    {
+        semantictiargsdone = 1u << (nest.sizeof * 8 - 1),
+        havetempdecl = semantictiargsdone >> 1,
+        gagged = semantictiargsdone >> 2
+    }
+
+    final @safe @property pure nothrow @nogc
+    {
+        /// has semanticTiargs() been done?
+        bool semantictiargsdone() const { return (nest & Flag.semantictiargsdone) != 0; }
+        void semantictiargsdone(bool x)
+        {
+            if (x) nest |= Flag.semantictiargsdone;
```

Comment:
This won't work, `nest` is written in other places as well? Why not add a dedicated `ubyte flags` (the required space is currently wasted as padding anyway)

Reply:
I think I should have mentioned - all code I could find that uses nest only does ++ and -- on it. A previous version had an explicit `flags` var but I just wanted to make things as compact as possible.

Reply:
@MoonlightSentinel see changes in the last commit that protects `nest` against accidental obliteration.

Reply:
TBH this just adds unecessary complication and worsens the debugging expirience ....

Reply:
I empathize. Problem at Symmetry is a 32 GB machine with 8 GB swap file can barely build the system. Just barely - I must close all JVM apps and sometimes Thunderbird (why does that eat so much RAM?) just to build. The size of the system ain't going down.

Even if this one reduction does not move the needle (what with 16 bytes alignment in the allocator and all), cutting down the size of `TemplateInstance` is high yield because there's so many of them. So I see this PR more like a step of several rather than an end in its own. (FWIW if I change alignment in the allocator from 16 to 8 bytes, this PR alone already produces a 1% memory improvement.)

We could debate the merits of this that or the other approach, but at the end of the day reducing n = `instance_size_of_TemplateInstance * number_of_TemplateInstance_objects` is a step in the right direction.

Reply:
Why not move the `nest` member out of `TemplateInstance` then? It's only used inside `headergen.d` when generating `*.di` files.